File: android/Xtrade/app/src/main/java/nz/co/fendustries/xtrade/fragments/CalculateFragment.java
package nz.co.fendustries.xtrade.fragments;

import android.content.Context;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.InputMethodManager;
import android.widget.EditText;

import javax.inject.Inject;

import dagger.Lazy;
import nz.co.fendustries.xtrade.R;
import nz.co.fendustries.xtrade.XtradeApplication;
import nz.co.fendustries.xtrade.activities.HomeActivity;
import nz.co.fendustries.xtrade.adapters.ConvertedRatesRecyclerAdapter;
import nz.co.fendustries.xtrade.helpers.AndroidConstants;
import nz.co.fendustries.xtrade.interfaces.CalculateViewContract;
import nz.co.fendustries.xtrade.presenters.CalculatePresenter;

/**
 * Created by joshuafenemore on 21/10/16.
 */
public class CalculateFragment extends Fragment implements CalculateViewContract
{
    private EditText valueEditText;
    private RecyclerView ratesRecyclerView;
    private RecyclerView.LayoutManager ratesRecylerViewLayoutManager;
    private ConvertedRatesRecyclerAdapter ratesRecyclerAdapter;
    private String savedString = "";

    @Inject
    Lazy<CalculatePresenter> calculatePresenterLazy;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState)
    {
        super.onCreateView(inflater, container, savedInstanceState);
        ((XtradeApplication)this.getActivity().getApplication()).getXtradeComponent().inject(this);

        View view = inflater.inflate(R.layout.fragment_calculate, container, false);
        this.valueEditText = (EditText) view.findViewById(R.id.valueEditText);
        this.ratesRecyclerView = (RecyclerView) view.findViewById(R.id.calculatedRatesRecyclerView);
        this.ratesRecylerViewLayoutManager = new LinearLayoutManager(this.getActivity());
        this.ratesRecyclerView.setLayoutManager(this.ratesRecylerViewLayoutManager);

        if (savedInstanceState != null)
        {
            this.savedString = savedInstanceState.getString(AndroidConstants.RATE_ENTERED, "");
        }

        return view;
    }

    @Override
    public void onResume()
    {
        super.onResume();

        ((HomeActivity) this.getActivity()).setActionBarTitle(this.getString(R.string.calculateLabel));

        this.valueEditText.addTextChangedListener(this.textWatcher);

        this.calculatePresenterLazy.get().setCalculateViewContract(this, this.savedString);
    }

    @Override
    public void onPause()
    {
        super.onPause();

        InputMethodManager imm = (InputMethodManager) this.getActivity().getSystemService(Context.INPUT_METHOD_SERVICE);
        imm.hideSoftInputFromWindow(this.getActivity().getWindow().getDecorView().getWindowToken(), 0);

        this.ratesRecyclerAdapter = null;
        this.valueEditText.removeTextChangedListener(this.textWatcher);
        this.calculatePresenterLazy.get().setCalculateViewContract(null, null);
    }

    @Override
    public void onSaveInstanceState(Bundle outgoingState)
    {
        outgoingState.putString(AndroidConstants.RATE_ENTERED, this.valueEditText.getText().toString());
        super.onSaveInstanceState(outgoingState);
    }

    @Override
    public void updateView()
    {
        if (this.ratesRecyclerAdapter == null)
        {
            this.ratesRecyclerAdapter = new ConvertedRatesRecyclerAdapter((AppCompatActivity) this.getActivity(), this.calculatePresenterLazy.get().getConvertedRateViewModels());
            this.ratesRecyclerView.setAdapter(this.ratesRecyclerAdapter);
        }
        else
        {
            this.ratesRecyclerAdapter.notifyDataSetChanged();
        }

        this.valueEditText.removeTextChangedListener(this.textWatcher);
        this.valueEditText.setText(this.calculatePresenterLazy.get().getDollarValue());
        this.valueEditText.setSelection(this.calculatePresenterLazy.get().getDollarValue().length());
        this.valueEditText.addTextChangedListener(this.textWatcher);
    }

    private TextWatcher textWatcher = new TextWatcher()
    {
        @Override
        public void beforeTextChanged(CharSequence s, int start, int count, int after)
        {

        }

        @Override
        public void onTextChanged(CharSequence s, int start, int before, int count)
        {
            calculatePresenterLazy.get().updateData(s.toString());
        }

        @Override
        public void afterTextChanged(Editable s)
        {

        }
    };
}

File: xamarin/Xtrade/Xtrade.AndroidApp/fragments/CalculateFragment.cs
namespace Xtrade.AndroidApp.Fragments
{
    using System;
    using Activities;
    using Adapters;
    using Android.Content;
    using Android.OS;
    using Android.Support.V7.App;
    using Android.Support.V7.Widget;
    using Android.Views;
    using Android.Widget;
    using Shared.Interfaces.ViewModels;

    public class CalculateFragment : BaseFragment<IAllRatesViewModel>
    {
        private ImageView _forexFlagImageView;
        private RecyclerView _ratesRecyclerView;
        private RecyclerView.LayoutManager _ratesRecylerViewLayoutManager;
        private RatesRecyclerAdapter _ratesRecyclerAdapter;

        public override View OnCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState)
        {
            base.OnCreateView(inflater, container, savedInstanceState);

            View view = inflater.Inflate(Resource.Layout.fragment_calculate, container, false);
            this._forexFlagImageView = view.FindViewById<ImageView>(Resource.Id.forexFlagImageView);
            this._ratesRecyclerView = view.FindViewById<RecyclerView>(Resource.Id.ratesRecyclerView);
            this._ratesRecylerViewLayoutManager = new LinearLayoutManager(this.Activity);

            return view;
        }

        public override void OnResume()
        {
            base.OnResume();

            ((HomeActivity)this.Activity).SetActionBarTitle(this.GetString(Resource.String.allRatesLabel));

            this.ViewModel.OnViewModelDataChanged += this.ViewModelDataChanged;

            this.ViewModel.LoadData();
        }

        public override void OnPause()
        {
            base.OnPause();

            this.ViewModel.OnViewModelDataChanged -= this.ViewModelDataChanged;
        }

        private void UpdateViews()
        {
            if (this._ratesRecyclerAdapter == null)
            {
                this._ratesRecyclerAdapter = new RatesRecyclerAdapter((AppCompatActivity) this.Activity, this.ViewModel.AllRates, i =>
                {
                    Intent detailsIntent = new Intent(this.Activity, typeof (ExchangeRateDetailsActivity));
                    detailsIntent.PutExtra(Helpers.AndroidConstants.SelectedRateCode, this.ViewModel.AllRates[i].CurrencyCode);
                    this.StartActivity(detailsIntent);
                });

                this._ratesRecyclerView.SetAdapter(this._ratesRecyclerAdapter);
                this._ratesRecyclerView.SetLayoutManager(this._ratesRecylerViewLayoutManager);
            }
            else
            {
                this._ratesRecyclerAdapter.UpdateDataSet(this.ViewModel.AllRates);
            }
        }

        private void ViewModelDataChanged(object sender, EventArgs eventArgs)
        {
            this.Activity.RunOnUiThread(() =>
            {
                if (this.IsVisible)
                {
                    this.UpdateViews();
                }
            });
        }
    }
}